cpp/src/arrow/flight/sql/odbc/odbcabstraction/odbc_impl/odbc_statement.cc

```diff
@@ -0,0 +1,787 @@
+    case SQL_C_INTERVAL_YEAR_TO_MONTH:
+    case SQL_C_INTERVAL_MONTH:
+      return sizeof(SQL_INTERVAL_STRUCT);
+    default:
+      return record.m_length;
+  }
+}
+
+SQLSMALLINT getCTypeForSQLType(const DescriptorRecord& record) {
+  switch (record.m_conciseType) {
+    case SQL_CHAR:
+    case SQL_VARCHAR:
+    case SQL_LONGVARCHAR:
+      return SQL_C_CHAR;
+
+    case SQL_WCHAR:
+    case SQL_WVARCHAR:
+    case SQL_WLONGVARCHAR:
+      return SQL_C_WCHAR;
+
+    case SQL_BINARY:
+    case SQL_VARBINARY:
+    case SQL_LONGVARBINARY:
+      return SQL_C_BINARY;
+
+    case SQL_TINYINT:
+      return record.m_unsigned ? SQL_C_UTINYINT : SQL_C_STINYINT;
+
+    case SQL_SMALLINT:
+      return record.m_unsigned ? SQL_C_USHORT : SQL_C_SSHORT;
+
+    case SQL_INTEGER:
+      return record.m_unsigned ? SQL_C_ULONG : SQL_C_SLONG;
+
+    case SQL_BIGINT:
+      return record.m_unsigned ? SQL_C_UBIGINT : SQL_C_SBIGINT;
+
+    case SQL_REAL:
+      return SQL_C_FLOAT;
+
+    case SQL_FLOAT:
+    case SQL_DOUBLE:
+      return SQL_C_DOUBLE;
+
+    case SQL_DATE:
+    case SQL_TYPE_DATE:
+      return SQL_C_TYPE_DATE;
+
+    case SQL_TIME:
+    case SQL_TYPE_TIME:
+      return SQL_C_TYPE_TIME;
+
+    case SQL_TIMESTAMP:
+    case SQL_TYPE_TIMESTAMP:
+      return SQL_C_TYPE_TIMESTAMP;
+
+    case SQL_C_INTERVAL_DAY:
+      return SQL_INTERVAL_DAY;
+    case SQL_C_INTERVAL_DAY_TO_HOUR:
+      return SQL_INTERVAL_DAY_TO_HOUR;
+    case SQL_C_INTERVAL_DAY_TO_MINUTE:
+      return SQL_INTERVAL_DAY_TO_MINUTE;
+    case SQL_C_INTERVAL_DAY_TO_SECOND:
+      return SQL_INTERVAL_DAY_TO_SECOND;
+    case SQL_C_INTERVAL_HOUR:
+      return SQL_INTERVAL_HOUR;
+    case SQL_C_INTERVAL_HOUR_TO_MINUTE:
+      return SQL_INTERVAL_HOUR_TO_MINUTE;
+    case SQL_C_INTERVAL_HOUR_TO_SECOND:
+      return SQL_INTERVAL_HOUR_TO_SECOND;
+    case SQL_C_INTERVAL_MINUTE:
+      return SQL_INTERVAL_MINUTE;
+    case SQL_C_INTERVAL_MINUTE_TO_SECOND:
+      return SQL_INTERVAL_MINUTE_TO_SECOND;
+    case SQL_C_INTERVAL_SECOND:
+      return SQL_INTERVAL_SECOND;
+    case SQL_C_INTERVAL_YEAR:
+      return SQL_INTERVAL_YEAR;
+    case SQL_C_INTERVAL_YEAR_TO_MONTH:
+      return SQL_INTERVAL_YEAR_TO_MONTH;
+    case SQL_C_INTERVAL_MONTH:
+      return SQL_INTERVAL_MONTH;
+
+    default:
+      throw DriverException("Unknown SQL type: " + std::to_string(record.m_conciseType),
+                            "HY003");
+  }
+}
+
+void CopyAttribute(Statement& source, Statement& target,
+                   Statement::StatementAttributeId attributeId) {
+  auto optionalValue = source.GetAttribute(attributeId);
+  if (optionalValue) {
+    target.SetAttribute(attributeId, *optionalValue);
+  }
+}
+}  // namespace
+
+// Public
+// =========================================================================================
+ODBCStatement::ODBCStatement(
+    ODBCConnection& connection,
+    std::shared_ptr<driver::odbcabstraction::Statement> spiStatement)
+    : m_connection(connection),
```

Comment:
We generally don't use the `m_` prefix in this codebase. It should be `connection_` if it's private.